### ISSUE TITLE
Issues Caused by NonBoltOrderObserver

### ISSUE THREAD

**Issue by mindfriction** [BoltApp/bolt-magento2#790]
Hello,

We began experiencing a number of issues with Non-Bolt Orders. After an Non-Bolt order was placed via front/backend, the following error would occur:
Warning: Invalid argument supplied for foreach() in /chroot/home/xxxx/xxxx.com/html/vendor/magento/module-sales/Model/OrderRepository.php on line 186

This same error would occur when trying to open the order from the Sales Order Grid.

In addition, we could no longer pull orders using the REST Orders API due to 500 Internal Errors

After further investigation, all of the above errors were caused by malformed data that was set by the NonBoltOrderObserver (#716) in the additional_information column within the sales_order_payment table

For example, the data in additional_information for all non-bolt orders started to appear like this:
"{\"method_title\":\"PayPal\",\"avsPostalCodeResponseCode\":\"I\",\"avsStreetAddressResponseCode\":\"I\",\"cvvResponseCode\":\"I\",\"processorAuthorizationCode\":null,\"processorResponseCode\":\"1000\",\"processorResponseText\":\"Approved\",\"paymentId\":\"PAYID-XXX\",\"payerEmail\":\"email@address.org\",\"bolt_transaction_reference\":\"XXX-XXX-XXX\"}"

When it should look like this, instead:{"method_title":"PayPal","avsPostalCodeResponseCode":"I","avsStreetAddressResponseCode":"I","cvvResponseCode":"I","processorAuthorizationCode":null,"processorResponseCode":"1000","processorResponseText":"Approved","paymentId":"PAYID-XXX","payerEmail":"email@address.org","bolt_transaction_reference":"XXX-XXX-XXX"}

Some of your more recent code, specifically the new NonBoltOrderObserver is setting this additional data and storing it as an array via     

            $boltTransactionData = ['bolt_transaction_reference' => $response->reference];
            $payment->setAdditionalInformation(
                array_merge((array)$payment->getAdditionalInformation(), $boltTransactionData)
            );
            $payment->save();

There also appears to be a similar code within Helper/Order.php using array_merge((array)

Magento's own function for setAdditionalInformation() doesn't allow setting arrays
    /**
     * Additional information setter
     * Updates data inside the 'additional_information' array
     * Doesn't allow to set arrays
     *
     * @param string $key
     * @param mixed $value
     * @return $this
     * @throws \Magento\Framework\Exception\LocalizedException
     */
    public function setAdditionalInformation($key, $value)
    {
        if (is_object($value)) {
            throw new \Magento\Framework\Exception\LocalizedException(
                __('Payment transactions disallow storing objects.')
            );
        }
        $info = $this->_getData('additional_information');
        if (!$info) {
            $info = [];
        }
        $info[$key] = $value;
        return $this->setData('additional_information', $info);
    }

In order to resolve the errors on my store, I had to remove the nonBoltOrderObserver from the sales_order_place_after event to stop this from happening. In addition, spent a significant amount of time manually cleaning up all of the malformed data from hundreds of non-bolt orders that had the malformed data set by the observer. Otherwise we could not download orders to our Shipping platform via Rest Orders API

**Comment by adentong** [comment]
@nickfiacco-bolt @matt-thomason @vitaliyreznikov

**Comment by roopakv** [comment]
@mindfriction thank you for posting such a detailed issue. We apologize for the issues that this issue caused you. Would you be able to share your email so that we can get some more info from you?

you can also reach me at roopak[at]bolt.com

**Comment by roopakv** [comment]
@matt-thomason could you please update this issue with he guidelines we came up with, and then close it

**Comment by nickfiacco-bolt** [comment]
Hi @mindfriction, thank you for creating this issue.

The `array_merge` code you saw was added after we had already realized something was wrong, and this is actually the correct way of setting the `additional_information` field. The previous code was setting the field via: `setAdditionalInformation('bolt_transaction_reference', $reference)` which was incorrect.

The documentation that says "Doesn't allow to set arrays" is referring to the fact that the *values* cannot be arrays, although it allows passing an array to the function itself (if you call the Magento REST Orders API and you can see what the `additional_information` field looks like).

While we test this fix thoroughly, we have removed the Non-Bolt Orders Observer from event registration for the next version of the plugin, and will not be rolling it out until it is wrapped in a feature switch.

**Comment by nickfiacco-bolt** [comment]
Hi @mindfriction, thank you for reporting the issue and providing a detailed write-up!

We've removed the `NonBoltOrderObserver` from being registered for the `sales_order_place_after` event for the time being, and will not register it again until we have wrapped it in a feature switch and thoroughly investigated the cause of your issue.

I'm going to close the issue, let me know if you have any other questions.
